toolchain/semantics/semantics_context.cpp

```diff
@@ -569,6 +587,10 @@ auto SemanticsContext::ParamOrArgSave(bool for_args) -> void {
 
 auto SemanticsContext::CanonicalizeType(SemanticsNodeId node_id)
     -> SemanticsTypeId {
+  auto node = semantics_ir_->GetNode(node_id);
+  if (node.kind() == SemanticsNodeKind::StubReference) {
+    node_id = node.GetAsStubReference();
+  }
```

Comment:
Should this be:
```suggestion
  while (node.kind() == SemanticsNodeKind::StubReference) {
    node_id = node.GetAsStubReference();
    node = semantics_ir_->GetNode(node_id);
  }
```
or is it guaranteed that we don't form a `StubReference` to a `StubReference`?

Reply:
Thank you! That is a great catch! I can't think of cases where `StubReference` is formed for a `StubReference` . But @jonmeow can answer this better than me.

Reply:
Right now I would think it's a bug if we have a recursive stub reference. While it may turn out to be an issue in the future, my preference would be either to not handle it or add a CARBON_CHECK that a while loop isn't needed -- that way if we start doing it we can consider the necessity or if there's a more efficient approach (part of my concern is accidentally adding extra stub unwinding when we could unwind once).

Reply:
Per question, I am saying I don't think this should have a `while`. Either:

1) Do nothing (just don't support it).
2) Explicitly CARBON_CHECK for the case that a stub reference points at another. This is the case where a `while` would have had an effect, but explicitly fails instead of supporting.

In both cases, the idea is to not support it, so that if it's needed we're more likely to evaluate support based on need at that time.

Reply:
Thank you! I added a carbon check so that we can detect if there's any such case.